### PR TITLE
refactor(OverlayTrigger): convert to function component

### PR DESCRIPTION
**Pull request by muzakparov**
Related to #4194

### CODE REVIEW

**Comment by taion** [comment]
First, great job refactoring this away from being a class component!

Second, sorry, but unfortunately, this looks like it was a bug in #4194. The `<OverlayTrigger>` itself probably shouldn't accept a ref – the `<RefHolder>` itself is just so we can attach a ref to find a DOM node. It's otherwise uninteresting and there's no particular reason why a user would _want_ a ref there. If anything, the user would want a ref to the wrapped component – but he or she can just drop a ref on that component directly.

I don't think we want to change `<OverlayTrigger>` to require the rendered child to accept a ref.

So... I think it'd be justifiable to just drop the `useMergedRefs` thing and make this a non-ref-forwarding component, but still a functional component that uses hooks (as it is now), rather than a class component.

**Comment by muzakparov** [comment]
@taion thank you for detailed explanation! :) 
could you please help me with tests? Since component was refactored to react hooks, some of the test related to  react `state` fail:
```
expect(wrapper.state('show')).to.equal(...)  // - fails
```
etc

should I just remove them?

**Comment by taion** [comment]
Can you instead look for the `<Overlay>` component instance and assert on the value of its `show` prop as passed down? Should be the same thing.

**Comment by muzakparov** [comment]
@taion sorry, I am bad with test and I couldn't come up with a way to access `show` prop. `<Overlay>` component turned out to be a function Component and I guess it doesn't have an instance.

**Comment by muzakparov** [comment]
@taion removed `forwardRef` along with `useMergedRef`

File: src/OverlayTrigger.js [hunk 1]
```diff
@@ -1,8 +1,7 @@
 /* eslint-disable max-classes-per-file */
```
Comment (taion) [inline src/OverlayTrigger.js]:
```suggestion
/* eslint-disable max-classes-per-file */

```

File: src/OverlayTrigger.js [hunk 2]
```diff
@@ -82,189 +81,175 @@ const defaultProps = {
 };
 
 // eslint-disable-next-line react/no-multi-comp
```
Comment (taion) [inline src/OverlayTrigger.js]:
```suggestion
```

no longer need this pragma as we only have one class component now

File: src/OverlayTrigger.js [hunk 3]
```diff
@@ -82,189 +81,175 @@ const defaultProps = {
...
-        const target = this.getTarget();
-        if (!this.state.show || !target) return data;
-
-        const role = popper.getAttribute('role') || '';
-        if (popper.id && role.toLowerCase() === 'tooltip') {
-          target.setAttribute('aria-describedby', popper.id);
-        }
-        return data;
-      },
+const OverlayTrigger = ({
```
Comment (taion) [inline src/OverlayTrigger.js]: to match the other components, could we have this as:

```js
function OverlayTrigger({ ... })
```

i.e. we don't use arrow functions except with `React.forwardRef`

File: src/OverlayTrigger.js [hunk 4]
```diff
@@ -82,189 +81,175 @@ const defaultProps = {
...
+        onHide={handleHide}
+        target={getTarget}
+      >
+        {overlay}
+      </Overlay>
+    </>
+  );
+};
+
+OverlayTrigger.displayName = 'OverlayTrigger';
```
Comment (taion) [inline src/OverlayTrigger.js]:
```suggestion
```

no longer needed as this is not a `React.forwardRef` any more

File: test/OverlayTriggerSpec.js [hunk 5]
```diff
@@ -232,111 +232,112 @@ describe('<OverlayTrigger>', () => {
...
-        wrapper.find('button').simulate('click');
-
-        // Need to click this way for it to propagate to document element.
-        document.getElementById('replace-overlay').click();
-
-        wrapper.state('show').should.be.true;
-      });
-    });
-  });
+  // We need different tests that doesn't rely on state
```
Comment (bpas247) [inline test/OverlayTriggerSpec.js]: Any chance we can re-write these to get them to work? otherwise we should just remove them.

Comment (muzakparov) [reply test/OverlayTriggerSpec.js]: removing them for now

File: src/OverlayTrigger.js [hunk 6]
```diff
@@ -77,192 +100,168 @@ const propTypes = {
...
+  defaultShow,
+  delay: propsDelay,
+  ...props
+}) {
+  const triggerNodeRef = useRef(null);
+  const timeoutRef = useRef();
+  const hoverStateRef = useRef();
+  const [show, setShow] = useState(!!defaultShow);
+
+  useEffect(() => {
```
Comment (jquense) [inline src/OverlayTrigger.js]: `useWillUnmount()` simplifies this in restart/hooks

File: src/OverlayTrigger.js [hunk 7]
```diff
@@ -77,192 +100,168 @@ const propTypes = {
...
-    clearTimeout(this._timeout);
-  }
+  const child = React.Children.only(children);
+  const { onFocus, onBlur, onClick } = child.props;
 
-  getChildProps() {
-    return React.Children.only(this.props.children).props;
-  }
+  const getTarget = useCallback(
+    () => ReactDOM.findDOMNode(triggerNodeRef.current),
```
Comment (jquense) [inline src/OverlayTrigger.js]: we should consider dropping the `findDOMNode` before v1 like we did for dropdowns

File: src/OverlayTrigger.js [hunk 8]
```diff
@@ -77,192 +100,168 @@ const propTypes = {
...
+  trigger,
+  overlay,
+  children,
+  popperConfig = {},
+  defaultShow,
+  delay: propsDelay,
+  ...props
+}) {
+  const triggerNodeRef = useRef(null);
+  const timeoutRef = useRef();
```
Comment (jquense) [inline src/OverlayTrigger.js]: also in restart/hooks `useTimeout` simplifies the logic here a bit

File: src/OverlayTrigger.js [hunk 9]
```diff
@@ -77,192 +96,164 @@ const propTypes = {
...
+}) {
+  const triggerNodeRef = useRef(null);
+  const timeout = useTimeout();
+  const hoverStateRef = useRef();
+  const [show, setShow] = useState(!!defaultShow);
 
-  componentWillUnmount() {
-    clearTimeout(this._timeout);
-  }
+  useWillUnmount(() => timeout.clear());
```
Comment (taion) [inline src/OverlayTrigger.js]: oh, we can actually remove this entirely, if that's all we need from unmounting

`useTimeout` will clear itself on unmount

Comment (bpas247) [reply src/OverlayTrigger.js]: It even calls `useWillUnmount` internally.
https://github.com/react-restart/hooks/blob/master/src/useTimeout.ts#L41

File: src/OverlayTrigger.js (hunk 7 above)
Comment (bpas247) [reply src/OverlayTrigger.js]: @jquense would that be out of scope for this PR? Or should we do this in this PR?

Comment (taion) [reply src/OverlayTrigger.js]: not in scope for this PR, i don't think

File: src/OverlayTrigger.js (hunk 9 above)
Comment (muzakparov) [reply src/OverlayTrigger.js]: removed it, thanks!

**Comment by taion** [comment]
thanks!

**Pull request merged**